src/devices/sound/nds_sound.h

```diff
@@ -0,0 +1,203 @@
+
+class nds_sound_device : public device_t, public device_sound_interface, public device_memory_interface
+{
+public:
+	static constexpr feature_type imperfect_features() { return feature::SOUND; }
+
+	// construction/destruction
+	nds_sound_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
+
+	// channel registers
+	u32 channel_control_r(offs_t offset);
+	void channel_w(offs_t offset, u32 data, u32 mem_mask);
+
+	// capture registers
+	u8 capture_control_r(offs_t offset);
+	u32 capture_addr_r(offs_t offset);
+	void capture_control_w(offs_t offset, u8 data);
+	void capture_addrlen_w(offs_t offset, u32 data, u32 mem_mask);
+
+	// global registers
+	u32 control_r();
+	u32 bias_r();
+	void control_w(offs_t offset, u32 data, u32 mem_mask);
+	void bias_w(offs_t offset, u32 data, u32 mem_mask);
+
+	// register map
+	void amap(address_map &map);
+protected:
+	// device-level overrides
+	virtual void device_start() override;
+	virtual void device_reset() override;
+	virtual void device_clock_changed() override;
+
+	// device_sound_interface overrides
+	virtual void sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs) override;
+
+	// device_memory_interface configuration
+	virtual space_config_vector memory_space_config() const override;
+
+	address_space_config m_data_config;
+
+	memory_access<27, 2, 0, ENDIANNESS_LITTLE>::cache    m_cache;
+	memory_access<27, 2, 0, ENDIANNESS_LITTLE>::specific m_space;
+
+private:
+	// structs
+	static constexpr unsigned CLOCK_DIVIDE = 512; // for genetate sample rate, correct?
+	enum
+	{
+		STATE_ADPCM_LOAD = 0,
+		STATE_PRE_LOOP,
+		STATE_POST_LOOP
+	};
+
+	struct channel_t 
+	{
+		// inline constants
+		const u8 voldiv_shift[4] = {0, 1, 2, 4};
+
+		// control bits
+		inline int volume() const { return BIT(control, 0, 7); } // global volume
+		inline u32 voldiv() const { return voldiv_shift[BIT(control, 8, 2)]; } // volume shift
+		inline bool hold() const  { return BIT(control, 15); } // hold bit
+		inline u32 pan() const    { return BIT(control, 16, 7); } // panning (0...127, 0 = left, 127 = right, 64 = half)
+		inline u32 duty() const   { return BIT(control, 24, 3); } // PSG duty
+		inline u32 repeat() const { return BIT(control, 27, 2); } // Repeat mode (Manual, Loop infinitely, One-shot)
+		inline u32 format() const { return BIT(control, 29, 2); } // Sound Format (PCM8, PCM16, ADPCM, PSG/Noise when exists)
+		inline bool busy() const  { return BIT(control, 31); } // Busy flag
+
+		// calculated values
+		inline int lvol() const   { return (pan() == 0x7f) ? 0 : 128 - pan(); } // calculated left volume
+		inline int rvol() const   { return (pan() == 0x7f) ? 128 : pan(); } // calculated right volume
+
+		// calculated address
+		inline u32 addr() const { return (sourceaddr & ~3) + (cur_bitaddr >> 3) + (cur_state == STATE_POST_LOOP ? ((loopstart + cur_addr) << 2) : (cur_addr << 2)); }
```

Comment:
There’s no point using the `inline` keyword with member functions where the body is provided at declaration like this – it’s implied, and having it there makes the lines longer and more cluttered.